board.cpp

```diff
@@ -0,0 +1,190 @@
+	this->blocksPerRow = std::vector<int>(BOARD::COLUMN_SIZE, 0);
+	this->currentHeight = 0;
+	this->currentPiece = new Piece();
+	this->absPiecePositionCol = 0;
+	this->absPiecePositionRow = 0;
+	this->rotation = 0;
+}
+
+int& Board::operator[] (int rowCol){
+	return board[rowCol];
+}
+
+
+int& Board::operator[] (indices rowCol) {
+	return board[rowCol.i * 10 + rowCol.j];
+}
+
+inline int Board::getPieceCol(int i) {
+	return this->absPiecePositionCol + (*(this->currentPiece->relXPositions))[rotation * 4 + i];
+}
+
+inline int Board::getPieceRow(int i) {
+	return this->absPiecePositionRow + (*(this->currentPiece->relYPositions))[rotation * 4 + i];
+}
+
+inline int Board::getPieceRowCol(int i) {
+	return 10 * this->absPiecePositionRow + this->absPiecePositionCol + (*(this->currentPiece->relXPositions))[rotation * 4 + i] + 10 * (*(this->currentPiece->relYPositions))[rotation * 4 + i];
+}
+
+bool Board::checkPieceOverlaps(int offsetX, int offsetY) {
+	return (board[((getPieceRowCol(0) + offsetX + BOARD::ROW_SIZE * offsetY + BOARD::BOARD_SIZE) % BOARD::BOARD_SIZE)] != 0)
+	    || (board[((getPieceRowCol(1) + offsetX + BOARD::ROW_SIZE * offsetY + BOARD::BOARD_SIZE) % BOARD::BOARD_SIZE)] != 0)
+	    || (board[((getPieceRowCol(2) + offsetX + BOARD::ROW_SIZE * offsetY + BOARD::BOARD_SIZE) % BOARD::BOARD_SIZE)] != 0)
+	    || (board[((getPieceRowCol(3) + offsetX + BOARD::ROW_SIZE * offsetY + BOARD::BOARD_SIZE) % BOARD::BOARD_SIZE)] != 0);
+}
+
+inline bool Board::checkOutOfBoundsRight(int offsetX) {
+	return (getPieceCol(0) + offsetX > BOARD::RIGHT_BOUNDARY) || (getPieceCol(1) + offsetX > BOARD::RIGHT_BOUNDARY) || (getPieceCol(2) + offsetX > BOARD::RIGHT_BOUNDARY) || (getPieceCol(3) + offsetX > BOARD::RIGHT_BOUNDARY);
+}
+
+inline bool Board::checkOutOfBoundsLeft(int offsetX) {
+	return (getPieceCol(0) + offsetX < BOARD::LEFT_BOUNDARY) || (getPieceCol(1) + offsetX < BOARD::LEFT_BOUNDARY) || (getPieceCol(2) + offsetX < BOARD::LEFT_BOUNDARY) || (getPieceCol(3) + offsetX < BOARD::LEFT_BOUNDARY);
+}
+
+inline bool Board::checkOutOfBoundsBelow(int offsetY) {
+	return (getPieceRow(0) + offsetY < BOARD::BOTTOM_BOUNDARY) || (getPieceRow(1) + offsetY < BOARD::BOTTOM_BOUNDARY) || (getPieceRow(2) + offsetY < BOARD::BOTTOM_BOUNDARY) || (getPieceRow(3) + offsetY < BOARD::BOTTOM_BOUNDARY);
+}
+
+void Board::generateNewPiece() {
+	delete this->currentPiece;
+	this->currentPiece = this->pieceNames[rand() % PIECES::NUMBER_OF_PIECES]();
+	this->absPiecePositionCol = PIECES::INITIAL_ABS_POSITION_X;
+	this->absPiecePositionRow = PIECES::INITIAL_ABS_POSITION_Y;
+	this->rotation = PIECES::INITIAL_ROTATION;
+}
+
+void Board::generateNewPiece(Piece* piece) {
+	delete this->currentPiece;
+	this->currentPiece = piece;
+	this->absPiecePositionCol = PIECES::INITIAL_ABS_POSITION_X;
+	this->absPiecePositionRow = PIECES::INITIAL_ABS_POSITION_Y;
+	this->rotation = PIECES::INITIAL_ROTATION;
+
+}
+
+void Board::removeFilledRows(int rowStart) {
```

Comment:
good job on making this function more readable! Couple questions though, this algorithm seems to populate blocks from the last unfilled row assuming there are some filled rows in between. Is this a valid assumption when there are random gaps such that blocks from above rows don't all necessarily fall into the next available unfilled row? Also what happens if the unfilled row becomes filled as a result of removing some filled rows? Shouldn't you also be updating blocksPerRow as well?

Reply:
Was this addressed?

Reply:
Blocks stay in the same row so unfilled rows won't be filled. blockPerRow updates now.